File: backend/src/grimoire/api/auxiliary.py
"""Auxiliary-task REST routes.

Seven start endpoints (one per `TaskKind`), an accept and a discard
route, and an in-flight listing endpoint. The start endpoints kick off
streaming through the gateway and return the in-memory `AuxiliaryResult`
once the stream completes; live tokens flow over the campaign WebSocket
as ``aux_token`` / ``aux_complete`` / ``aux_error`` events.

See `docs/superpowers/specs/2026-05-19-auxiliary-tasks-design.md`.
"""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from grimoire.api.deps import OrchestratorDep
from grimoire.api.util import map_lookup_errors, to_payload
from grimoire.auxiliary.types import AuxiliaryTask, TaskKind
from grimoire.orchestrator.errors import (
    AuxiliaryAlreadyCommittedError,
    AuxiliaryNotFoundError,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/campaigns")


# --------------------------------------------------------------------------- #
# Request payloads
# --------------------------------------------------------------------------- #


class ImpersonatePCPayload(BaseModel):
    steering_hint: str | None = None


class RewritePostPayload(BaseModel):
    post_id: str
    edit_instruction: str


class ContinueAsPayload(BaseModel):
    character_ref: str
    target_post_id: str | None = None
    steering_hint: str | None = None


class WhatWouldXSayPayload(BaseModel):
    character_ref: str
    snippet: str


class BrainstormPayload(BaseModel):
    prompt: str = Field(..., alias="prompt")


class EditProsePayload(BaseModel):
    snippet: str
    edit_instruction: str


class TranslatePayload(BaseModel):
    snippet: str
    target_language: str


class AcceptAuxiliaryPayload(BaseModel):
    edited_text: str | None = None


# --------------------------------------------------------------------------- #
# Helpers
# --------------------------------------------------------------------------- #


async def _run(
    orchestrator: Any,
    campaign_id: str,
    task: AuxiliaryTask,
) -> dict[str, Any]:
    try:
        result = await orchestrator.run_auxiliary_task(
            campaign_id=campaign_id, task=task
        )
    except Exception as exc:
        raise map_lookup_errors(exc) from exc
    return _result_payload(result)


def _result_payload(result: Any) -> dict[str, Any]:
    return {
        "id": result.id,
        "kind": result.task.kind.value,
        "text": result.text,
        "model_used": result.model_used,
        "tokens": result.tokens,
        "pending_commit_action": result.pending_commit_action.value,
        "warnings": list(result.warnings),
        "completed_at": result.completed_at.isoformat(),
    }


# --------------------------------------------------------------------------- #
# Start endpoints (7)
# --------------------------------------------------------------------------- #


@router.post("/{campaign_id}/auxiliary/impersonate-pc")
async def impersonate_pc(
    campaign_id: str,
    payload: ImpersonatePCPayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.IMPERSONATE_PC,
        steering_hint=payload.steering_hint,
    )
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/rewrite-post")
async def rewrite_post(
    campaign_id: str,
    payload: RewritePostPayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.REWRITE_POST,
        target_post_id=payload.post_id,
        edit_instruction=payload.edit_instruction,
    )
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/continue-as")
async def continue_as(
    campaign_id: str,
    payload: ContinueAsPayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.CONTINUE_AS,
        target_character_ref=payload.character_ref,
        target_post_id=payload.target_post_id,
        steering_hint=payload.steering_hint,
    )
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/what-would-x-say")
async def what_would_x_say(
    campaign_id: str,
    payload: WhatWouldXSayPayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.WHAT_WOULD_X_SAY,
        target_character_ref=payload.character_ref,
        snippet=payload.snippet,
    )
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/brainstorm")
async def brainstorm(
    campaign_id: str,
    payload: BrainstormPayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(kind=TaskKind.BRAINSTORM, snippet=payload.prompt)
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/edit-prose")
async def edit_prose(
    campaign_id: str,
    payload: EditProsePayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.EDIT_PROSE,
        snippet=payload.snippet,
        edit_instruction=payload.edit_instruction,
    )
    return await _run(orchestrator, campaign_id, task)


@router.post("/{campaign_id}/auxiliary/translate")
async def translate(
    campaign_id: str,
    payload: TranslatePayload,
    orchestrator: OrchestratorDep,
) -> Any:
    task = AuxiliaryTask(
        kind=TaskKind.TRANSLATE,
        snippet=payload.snippet,
        target_language=payload.target_language,
    )
    return await _run(orchestrator, campaign_id, task)


# --------------------------------------------------------------------------- #
# Accept / Discard
# --------------------------------------------------------------------------- #


@router.post("/{campaign_id}/auxiliary/{result_id}/accept")
async def accept_auxiliary(
    campaign_id: str,
    result_id: str,
    orchestrator: OrchestratorDep,
    payload: AcceptAuxiliaryPayload | None = None,
) -> Any:
    edited_text = payload.edited_text if payload else None
    try:
        out = await orchestrator.accept_auxiliary(
            campaign_id, result_id, edited_text=edited_text
        )
    except AuxiliaryNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except AuxiliaryAlreadyCommittedError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except Exception as exc:
        raise map_lookup_errors(exc) from exc
    return to_payload(out)


@router.post("/{campaign_id}/auxiliary/{result_id}/discard")
async def discard_auxiliary(
    campaign_id: str,
    result_id: str,
    orchestrator: OrchestratorDep,
) -> Any:
    discarded = await orchestrator.discard_auxiliary(result_id)
    if not discarded:
        raise HTTPException(status_code=404, detail=f"auxiliary not found: {result_id}")
    return {"discarded": True, "result_id": result_id}


# --------------------------------------------------------------------------- #
# In-flight listing
# --------------------------------------------------------------------------- #


@router.get("/{campaign_id}/auxiliary/in-flight")
async def list_in_flight(
    campaign_id: str,
    orchestrator: OrchestratorDep,
) -> Any:
    # All in-flight aux results are listed; the API doesn't currently
    # tag results with a campaign id, so the response is global. Frontend
    # filters by id pattern when needed.
    results = list(orchestrator._inflight_aux.values())
    return [_result_payload(r) for r in results]


__all__ = ["router"]

File: frontend/src/api/auxiliary.ts
/**
 * Auxiliary-task REST client.
 *
 * Auxiliary tasks (drafts, rewrites, brainstorms, etc.) are non-canonical
 * model calls — they never mutate scene state until the user explicitly
 * accepts a result. The backend parks each result in an in-memory slot
 * keyed by the returned `id`; accept commits it via the appropriate
 * canonical path, discard throws it away.
 */

import { api } from "./client";

export type AuxiliaryKind =
  | "impersonate_pc"
  | "rewrite_post"
  | "continue_as"
  | "what_would_x_say"
  | "brainstorm"
  | "edit_prose"
  | "translate";

export type AuxiliaryCommitAction =
  | "submit_post"
  | "replace_post"
  | "append_post"
  | "copy"
  | "replace_draft";

export interface AuxiliaryResult {
  id: string;
  kind: AuxiliaryKind;
  text: string;
  model_used: string;
  tokens: number;
  pending_commit_action: AuxiliaryCommitAction;
  warnings: string[];
  completed_at: string;
}

export interface AcceptAuxiliaryResponse {
  committed: boolean;
  action: AuxiliaryCommitAction;
  result_id: string;
  text?: string;
  post_id?: string;
  alternate_id?: string;
  cascaded_replace?: boolean;
  turn_id?: string;
}

const enc = encodeURIComponent;
const base = (campaignId: string) => `/api/campaigns/${enc(campaignId)}/auxiliary`;

export const auxiliaryApi = {
  impersonatePC: (campaignId: string, steeringHint?: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/impersonate-pc`, {
      steering_hint: steeringHint,
    }),

  rewritePost: (campaignId: string, postId: string, editInstruction: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/rewrite-post`, {
      post_id: postId,
      edit_instruction: editInstruction,
    }),

  continueAs: (
    campaignId: string,
    characterRef: string,
    targetPostId?: string,
    steeringHint?: string,
  ) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/continue-as`, {
      character_ref: characterRef,
      target_post_id: targetPostId,
      steering_hint: steeringHint,
    }),

  whatWouldXSay: (campaignId: string, characterRef: string, snippet: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/what-would-x-say`, {
      character_ref: characterRef,
      snippet,
    }),

  brainstorm: (campaignId: string, prompt: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/brainstorm`, { prompt }),

  editProse: (campaignId: string, snippet: string, editInstruction: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/edit-prose`, {
      snippet,
      edit_instruction: editInstruction,
    }),

  translate: (campaignId: string, snippet: string, targetLanguage: string) =>
    api.post<AuxiliaryResult>(`${base(campaignId)}/translate`, {
      snippet,
      target_language: targetLanguage,
    }),

  accept: (campaignId: string, resultId: string, editedText?: string) =>
    api.post<AcceptAuxiliaryResponse>(`${base(campaignId)}/${enc(resultId)}/accept`, {
      edited_text: editedText,
    }),

  discard: (campaignId: string, resultId: string) =>
    api.post<{ discarded: boolean; result_id: string }>(
      `${base(campaignId)}/${enc(resultId)}/discard`,
    ),

  inFlight: (campaignId: string) => api.get<AuxiliaryResult[]>(`${base(campaignId)}/in-flight`),
};
